components/Buttons/src/MDCFloatingButton+Animation.m

```diff
@@ -0,0 +1,222 @@
+    if (completion) {
+      completion();
+    }
+    NSLog(@"%@", [self.layer animationKeys]);
+  };
+
+  [self.layer removeAllAnimations];
+  [self.imageView.layer removeAllAnimations];
+
+  if (animated) {
+    [CATransaction begin];
+    [CATransaction setDisableActions:YES];
+    [CATransaction setCompletionBlock:exitActions];
+
+    CABasicAnimation *overallScaleAnimation = [CABasicAnimation animationWithKeyPath:@"transform"];
+    overallScaleAnimation.toValue = [NSValue valueWithCATransform3D:scaleTransform];
+    overallScaleAnimation.timingFunction =
+        [[CAMediaTimingFunction alloc] initWithControlPoints:0.4f:0.0f:1.0f:1.0f];
+    overallScaleAnimation.fillMode = kCAFillModeForwards;
+    overallScaleAnimation.removedOnCompletion = NO;
+    overallScaleAnimation.duration = kMDCFloatingButtonExitDuration;
+#if TARGET_IPHONE_SIMULATOR
+    overallScaleAnimation.duration *= [self fab_dragCoefficient];
+#endif
+    [self.layer addAnimation:overallScaleAnimation forKey:@"kMDCFloatingButtontransformexit"];
+
+    CABasicAnimation *iconScaleAnimation = [CABasicAnimation animationWithKeyPath:@"transform"];
+    iconScaleAnimation.toValue = [NSValue valueWithCATransform3D:scaleTransform];
+    iconScaleAnimation.timingFunction =
+        [[CAMediaTimingFunction alloc] initWithControlPoints:0.0f:0.0f:0.2f:1.0f];
+    iconScaleAnimation.fillMode = kCAFillModeForwards;
+    iconScaleAnimation.removedOnCompletion = NO;
+    iconScaleAnimation.duration = kMDCFloatingButtonExitIconDuration;
+    iconScaleAnimation.beginTime = CACurrentMediaTime() + kMDCFloatingButtonExitIconOffset;
+#if TARGET_IPHONE_SIMULATOR
+    iconScaleAnimation.duration *= [self fab_dragCoefficient];
+    iconScaleAnimation.beginTime =
+        CACurrentMediaTime() + (kMDCFloatingButtonExitIconOffset * [self fab_dragCoefficient]);
+#endif
+    [self.imageView.layer addAnimation:iconScaleAnimation
+                                forKey:@"kMDCFloatingButtontransformexit"];
+
+    CABasicAnimation *opacityAnimation = [CABasicAnimation animationWithKeyPath:@"opacity"];
+    opacityAnimation.toValue = [NSNumber numberWithFloat:0];
+    opacityAnimation.timingFunction =
+        [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear];
+    opacityAnimation.fillMode = kCAFillModeForwards;
+    opacityAnimation.removedOnCompletion = NO;
+    opacityAnimation.duration = kMDCFloatingButtonOpacityDuration;
+    opacityAnimation.beginTime = CACurrentMediaTime() + kMDCFloatingButtonOpacityExitOffset;
+#if TARGET_IPHONE_SIMULATOR
+    opacityAnimation.duration *= [self fab_dragCoefficient];
+    opacityAnimation.beginTime =
+        CACurrentMediaTime() + (kMDCFloatingButtonOpacityExitOffset * [self fab_dragCoefficient]);
+#endif
+    [self.layer addAnimation:opacityAnimation forKey:@"kMDCFloatingButtonopacityexit"];
+
+    [CATransaction commit];
+  } else {
+    exitActions();
+  }
+}
+
+#pragma mark - UIView overrides to handle non-identity transforms
```

Comment:
I've tried adding a container view, but MDCButton uses a shadowLayer so we can't swap them out.  It might be possible to refactor MDCButton to have an MDCButtonContainerView with a shadowLayer, then have all of the subviews in there.  The button view itself would just be an empty/transparent rectangle.  I can spend a day on that approach to see how far I can get if you prefer.